kms-provider-kroxylicious-inmemory/src/main/java/io/kroxylicious/kms/provider/kroxylicious/inmemory/InMemoryKms.java

```diff
@@ -0,0 +1,230 @@
+/*
+ * Copyright Kroxylicious Authors.
+ *
+ * Licensed under the Apache Software License version 2.0, available at http://www.apache.org/licenses/LICENSE-2.0
+ */
+
+package io.kroxylicious.kms.provider.kroxylicious.inmemory;
+
+import java.security.GeneralSecurityException;
+import java.security.InvalidKeyException;
+import java.security.NoSuchAlgorithmException;
+import java.security.SecureRandom;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.UUID;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.function.Supplier;
+
+import javax.crypto.Cipher;
+import javax.crypto.IllegalBlockSizeException;
+import javax.crypto.KeyGenerator;
+import javax.crypto.SecretKey;
+import javax.crypto.spec.GCMParameterSpec;
+
+import io.kroxylicious.kms.service.DekPair;
+import io.kroxylicious.kms.service.Kms;
+import io.kroxylicious.kms.service.KmsException;
+import io.kroxylicious.kms.service.Serde;
+import io.kroxylicious.kms.service.UnknownAliasException;
+import io.kroxylicious.kms.service.UnknownKeyException;
+
+import edu.umd.cs.findbugs.annotations.NonNull;
+
+/**
+ * An in-memory KMS to be used only for testing.
+ * Note that this exposes public methods that are not part of the {@link Kms} interface which are used for those
+ * KMS operations which are outside the scope of Kroxy itself (such as key provisioning).
+ */
+public class InMemoryKms implements
+        Kms<UUID, InMemoryEdek> {
+
+    private static final String WRAP_ALGO = "AES_256/GCM/NoPadding";
+    public static final String KEY_ALGO = "AES";
+    private final Map<UUID, SecretKey> keys;
+    private final KeyGenerator aes;
+    private final int numIvBytes;
+    private final int numAuthBits;
+    private final SecureRandom secureRandom;
+    private final Map<String, UUID> aliases;
+
+    private final AtomicInteger numDeksGenerated;
+
+    InMemoryKms(int numIvBytes, int numAuthBits,
+                Map<UUID, SecretKey> keys,
+                Map<String, UUID> aliases,
+                AtomicInteger numDeksGenerated) {
```

Comment:
Because we want an `InMemoryKms` to function like a client of a true KMS service. So the lifetime of the keys and aliases (aka "the KMS state") is not the same as the lifetime of the KMS client itself. I.e. I should be able to instantiate two `InMemoryKms` instances which share the same KMS state, just as I can instantiate two AWS KMS client instances and they observe the same keys.

Note that the `UnitTestingKmsService` and `IntegrationTestingKmsService` differ in only in how they treat the lifetime of the KMS state. For unit testing it's fine for the KMS state to be coupled to an instance of a `KmsService`, because we can easily obtain `KmsService` instances as required for each test. For the integration tests the proxy itself obtains a `KmsService` instances when it builds the filter factory, but we want the test code to have a `KmsService` instances sharing the same state, so we use a `static` for the shared state.

This is slightly more flexible that the approach @k-wall was originally using, where he configured the in-memory KMS with keys and aliases ahead of time in the integration tests. It means we are able to integration test scenarios like an alias being changed during the lifetime of a filter.